Graph.java

```diff
@@ -0,0 +1,166 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+
+/**
+ * Created by basescu on 06.06.17.
+ */
+
+class Node {
+    public int key;     // letter index in initial matrix: for M[i][j], we have pos = i * cols + j
+    public char value;
+
+    public Node(int key, char value) {
+        this.key = key;
+        this.value = value;
+    }
+
+    public String toString() {
+        return "key = " + key + " value = " + value;
+    }
+}
+
+/*
+Each letter is a node. The graph has edges for each diagonal & straight traversal.
+Finding words means a graph traversal with word prunning. I chose DFS to
+potentially find complete words faster.
+ */
+public class Graph {
+
+    public final HashMap<Integer, Node> intToNode;
+    public final HashMap<Node, LinkedList<Node>> edges;
```

Comment:
Potential problem: `Node` does not have stable (as in: returning the same value for node objects mapped to the same integer identifier), so you rely on "pointer" to the object. It works here, but if someone changes the code a bit, they might get into some issues.
To avoid such problems one either defines `hashCode` and `equals` methods in the object used as key, or uses things like [AutoValue](https://github.com/google/auto/blob/master/value/userguide/index.md) to generate these automatically.
Alternatively you could use `Integer` (as defined in the map above) as a key.

That's from the code behavior perspective. From implementation perspective, if you need to use `Map<SomeObject, Something>`, it is usually a strong signal that `Something` should be simply a member of `SomeObject`.